### PR TITLE
core,eth: avoid unnecessary receipt processing for log filtering

### PR DESCRIPTION
**Pull request by s1na**
When filtering logs, after a matching block has been found a slim version of the receipts is fetched from the database. These slim receipts have to be processed to fill in some metadata. Specifically a bloom filter has to be created which adds some unnecessary overhead because this bloom is not used for filtering purposes.

In this PR we add some new methods that only try to process the underlying logs of a receipt and not much more. Note that this also means we lose the ability to cache the receipts in Blockchain.

In a very simple test (fetching `Transfer` events of an erc20 token for a few thousand blocks) the runtime went from ~12s to ~7s on my machine.

### CODE REVIEW

File: core/rawdb/accessors_chain_test.go [hunk 1]
```diff
@@ -432,6 +432,108 @@ func checkReceiptsRLP(have, want types.Receipts) error {
...
+		t.Fatalf("non existent receipts returned: %v", rs)
+	}
+	// Insert the body that corresponds to the receipts
+	WriteBody(db, hash, 0, body)
+
+	// Insert the receipt slice into the database and check presence
+	WriteReceipts(db, hash, 0, receipts)
+
+	logs := ReadLogs(db, hash, 0)
+	if logs == nil || len(logs) == 0 {
```
Comment (holiman) [inline core/rawdb/accessors_chain_test.go:509]:
```suggestion
	if len(logs) == 0 {
```
^ that is sufficient

File: core/rawdb/accessors_chain_test.go [hunk 2]
```diff
@@ -432,6 +432,108 @@ func checkReceiptsRLP(have, want types.Receipts) error {
...
+		t.Fatalf("no logs returned")
+	}
+	if !(len(logs) == 2 && len(logs[0]) == 2 && len(logs[1]) == 2) {
+		t.Fatalf("unexpected number of logs returned")
+	}
```
Comment (holiman) [inline core/rawdb/accessors_chain_test.go:514]:
```suggestion
	if have, want := len(logs), 2; have != want {
		t.Fatalf("unexpected number of logs returned, have %d want %d")
	}
	if have, want := len(logs[0]), 2; have != want {
		t.Fatalf("unexpected number of logs[0] returned, have %d want %d")
	}
	if have, want := len(logs[1]), 2; have != want {
		t.Fatalf("unexpected number of logs[1] returned, have %d want %d")
	}

```
That may be a bit more verbose, but the "have, want" pattern is pretty common

**Review by holiman — COMMENTED** [review]
I don't see where you actually make use of this optimization?

File: eth/api_backend.go [hunk 3]
```diff
@@ -177,13 +177,14 @@ func (b *EthAPIBackend) GetReceipts(ctx context.Context, hash common.Hash) (type
...
-		return nil, nil
+	db := b.eth.ChainDb()
+	number := rawdb.ReadHeaderNumber(db, hash)
+	if number == nil {
+		return nil, errors.New("failed to get block number from hash")
 	}
-	logs := make([][]*types.Log, len(receipts))
-	for i, receipt := range receipts {
-		logs[i] = receipt.Logs
+	logs := rawdb.ReadLogs(db, hash, *number)
```
Comment (s1na) [inline eth/api_backend.go:185]: @holiman Changed the backend's `GetLogs` here to utilize the new methods. `GetLogs` is in turn called by `Filter.checkMatches`

File: core/rawdb/accessors_chain_test.go (hunk 1 above)
Comment (s1na) [reply core/rawdb/accessors_chain_test.go]: Nice didn't know you could do `len` on a nil slice

File: core/rawdb/accessors_chain.go [hunk 4]
```diff
@@ -607,6 +607,37 @@ func ReadReceipts(db ethdb.Reader, hash common.Hash, number uint64, config *para
 	return receipts
 }
 
+// ReadLogs retrieves the logs for all transactions in a block. The log fields
+// are populated with metadata. In case the receipts or the block body
+// are not found, a nil is returned.
+func ReadLogs(db ethdb.Reader, hash common.Hash, number uint64) [][]*types.Log {
```
Comment (holiman) [inline core/rawdb/accessors_chain.go:613]: Triage discussion, something like 
```suggestion
type LogFilterFn func(receipts []*types.ReceiptLogs) bool

func ReadLogs(db ethdb.Reader, hash common.Hash, number uint64, filterFn LogFilterFn) [][]*types.Log {
```
If the filter function returns `false`, we can omit the ReadBody and deriving the fields.

**Pull request closed without merging**